### PR TITLE
GPU Memory Pool, strings, and minor fixes

### PR DESCRIPTION
**Pull request by tbennun**
- [x] GPU memory pool allocation and release
- [x] Marshalling strings as arguments
- [x] Support `len` and `sum` on numpy arrays
- [x] Support degenerate (0-axis) reductions
- [x] TransientReuse pass doesn't erase arrays on its own
- [x] Further fixes to problematic view allocation
- [x] Fix bug in loop detection for loop2map
- [x] Remove mandatory cuda check

### CODE REVIEW

File: dace/transformation/passes/transient_reuse.py [hunk 1]
```diff
@@ -151,17 +151,6 @@ def apply_pass(self, sdfg: SDFG, _) -> Optional[Set[str]]:
                                 if edge.data.data == old:
                                     edge.data.data = new
 
-        # clean up the arrays
-        for a in list(sdfg.arrays):
```
Comment (alexnick83) [inline dace/transformation/passes/transient_reuse.py:155]: I guess we have an API call for this piece of code, but why was it removed completely?

**Review by alexnick83 — APPROVED** [review]
LGTM. I have some questions, but nothing blocking.

File: dace/codegen/targets/cuda.py [hunk 2]
```diff
@@ -194,6 +206,76 @@ def preprocess(self, sdfg: SDFG) -> None:
...
+        reachability = access_nodes = None
+        for sdfg in top_sdfg.all_sdfgs_recursive():
+            # Skip SDFGs without memory pool hints
+            pooled = set(aname for aname, arr in sdfg.arrays.items()
+                         if getattr(arr, 'pool', False) is True and arr.transient)
+            if not pooled:
+                continue
+            self.has_pool = True
+            if self.backend != 'cuda':
+                raise ValueError(f'Backend "{self.backend}" does not support the memory pool allocation hint')
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/codegen/targets/cuda.py:227]: Yes, however: AMD will need a runtime (.cuh file) of its own. CPU will probably draw from the same code or use existing widely-used mempools.

File: dace/codegen/targets/cpu.py [hunk 3]
```diff
@@ -1108,13 +1108,14 @@ def memlet_definition(self,
             else:
                 conntype = conntype.dtype
 
-        is_scalar = not isinstance(conntype, dtypes.pointer)
+        desc = sdfg.arrays[memlet.data]
+
+        is_scalar = not isinstance(conntype, dtypes.pointer) or desc.dtype == conntype
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/codegen/targets/cpu.py:1113]: Yes

File: dace/codegen/targets/cpp.py [hunk 4]
```diff
@@ -1337,15 +1348,54 @@ def presynchronize_streams(sdfg, dfg, state_id, node, callsite_stream):
...
+        terminators.remove(scope_exit)
+        if len(terminators) == 0:
+            desc = sd.arrays[name]
+            ptrname = ptr(name, desc, sd, codegen._frame)
+            if isinstance(desc, data.Array) and desc.start_offset != 0:
+                ptrname = f'({ptrname} - {sym2cpp(desc.start_offset)})'
+            if Config.get_bool('compiler', 'cuda', 'syncdebug'):
+                callsite_stream.write(f'DACE_CUDA_CHECK({backend}FreeAsync({ptrname}, {cudastream}));\n', sdfg, state_id, scope_exit)
+                callsite_stream.write(f'DACE_CUDA_CHECK({backend}DeviceSynchronize());')
+            else:
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/codegen/targets/cpp.py:1383]: the lines here refer to the syncdebug. It's like any config entry that we don't set by default but codecov is very strict about complaining.

**Pull request merged**